### PR TITLE
DashR Oil and Gas Ternary

### PR DESCRIPTION
**Pull request by KPhans**
Issue for app: #[issue number here]

# App pull request
- [ ] This is a new app
- [ ] I am improving an existing app (redesigns/code "makeovers")

## About

- Playground deployment URL (new version):
- Current gallery app URL: (delete this line if inapplicable)
- Python app repository link: (delete this line if you are working on a Python app)

## Workflow

- [ ] I have created a branch in the appropriate monorepo, and the
  elements necessary for successful deployment are in place.
- [ ] If the app is a redesigned and/or restyled version of an
  existing gallery app, I've summarized the changes requested in the
  appropriate Streambed issue and confirm that they have been applied.
- [ ] If the app is on the Dash Gallery portal, I have added a link to
  the GitHub repository for the source code in the portal description.
- [ ] If the app is a reimplementation of a Python gallery app for the
  DashR gallery, the app in this PR mimics, as closely as possible,
  the style and functionality of the existing app.=
- [ ] I have removed *all Google Analytics code* from the app's
  `assets/` folder.

## The pre-review review

I have addressed all of the following questions:

- [ ] Does everything in my code serve some purpose? (I have removed
  any dead and/or irrelevant code.)
- [ ] Does everything in my code have a clear purpose? (My code is
  readable and, where it isn't, it has been commented appropriately.)]
- [ ] Am I reinventing the wheel? (I have used appropriate packages to
  lessen the volume of code that needs to be maintained.)

### CODE REVIEW

**Comment by KPhans** [comment]
@rpkyle PR for Ternary app to be merged into master!

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+  )
+  
+  return(list(data = data, layout = layout))
+  
+}
+
+
+
+generate_ternary_map = function(dff, selected_data, contour_visible, marker_visible){
+  contour_traces = list()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: I'm probably in an extreme minority here, but I actually consider defining empty lists prior to assigning into their elements (e.g. while looping) to be a bit of an anti-pattern in R-- there are definitely contexts where it makes sense, but often you can either use `lapply` to generate a list of the appropriate dimensions, or manually define the list structure up front and then "fill in" as you go. 

I wouldn't worry about this for now, but I think it's good practice to keep in mind for the future.

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+      line=list(color='#444', width=0.5),
+      fill='toself',
+      fillcolor=ternary_color[j][[1]],
+      opacity=0.8,
+      hoverinfo='all',
+      showlegend=FALSE,
+      visible=contour_visible
+    )
+    j = j+1
+    #contour_traces = append(contour_traces,as.list(trace))
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: :hocho: commented out line

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+      b=dff[dff$fm_name == key,'Carbonate'],
+      c=dff[dff$fm_name == key,'Clay'],
+      mode='markers',
+      hovertemplate = hovertemplate,
+      showlegend=TRUE,
+      marker=list('color' = colormap[[key]], 'size' = 8, 'line' = list('color' = '#000000', 'width'= 0.2)),
+      selectedpoints=select_indices,
+      visible=marker_visible
+    )
+    #data_traces = append(data_traces,new_data_trace)
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: :hocho: commented out line

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+}
+
+generate_formation_bar = function(dff, selected_data){
+  
+  formations = list(unique(dff['fm_name']))
+  xform <- list(categoryorder = "array", 
+                categoryarray = formations[[1]]$fm_name,
+                tickangle=-45, title="Formations")
+  data = list()
+  countt = list()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: I'd suggest renaming `countt` to something more descriptive if possible.

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+  color = list()
+  
+  for (i in formations[[1]]$fm_name) {
+    counttt = nrow(dff[dff$fm_name == i,])
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: Same with `counttt` here, and any other variation in which ≥ 2 `t`'s are added 🙂

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+)
+))
+app$callback(output = list(id = "form-by-bar", property = "figure"),
+             params = list(
+               input("well-map", "selectedData"),
+               input("ternary-map", "selectedData"),
+               input("operator-select", "value")),
+             function(map_selected_data, tern_selected_data, op_select){
+               dff = df[df$op %in% op_select,]
+               #browser()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: :hocho: `#browser()`

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+             params = list(
+               input("well-map", "selectedData"),
+               input("form-by-bar", "selectedData"),
+               input("form-by-bar", "clickData"),
+               input("operator-select", "value"),
+               input("ternary-layer-select", "value"),
+               state("ternary-map", "figure")),
+             function(map_selected_data, bar_selected_data, bar_click_data, op_select,
+                      layer_select, curr_fig){
+               #browser()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: 💣 `#browser()`

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+             })
+
+app$callback(output = list(id = "well-map", property = "figure"),
+             params = list(input("ternary-map", "selectedData"),
+                           input("form-by-bar", "selectedData"),
+                           input("form-by-bar", "clickData"),
+                           input("operator-select", "value"),
+                           input("mapbox-view-selector", "value")),
+             function(tern_selected_data, bar_selected_data, bar_click_data, op_select, mapbox_view){
+               #browser()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: `#browser()` 🔫

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+               return(generate_well_map(dff, processed_data, mapbox_view))
+             })
+
+app$callback(output = list(id = "production-fig", property = "figure"),
+             params = list(input("well-map", "selectedData"),
+                           input("ternary-map", "selectedData"),
+                           input("form-by-bar", "selectedData"),
+                           input("operator-select", "value")),
+             function(map_select, tern_select, bar_select, op_select){
+               #browser()
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: 💥 ☠️ ⚛️ `#browser()`

File: apps/dashr-oil-and-gas-ternary/app.R [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+                                                                     'fm_name']))
+                   }
+                 } else{
+                   processed_data = processed_data_init
+                 }
+               } else if(prop_id == "ternary-map" & prop_type == "selectedData"){
+                 if(length(tern_select) > 0){
+                   processed_data = list('well_id' = list(), 'formation' = list())
+                   for (point in tern_select[['points']]) {
+                     #if('customdata' %in% point){
```
Comment (rpkyle) [inline apps/dashr-oil-and-gas-ternary/app.R]: :hocho: commented out line

**Comment by KPhans** [comment]
@rpkyle i believe everything is corrected. As per the css issue, i did not notice it when i ran it on my system.

File: apps/dashr-oil-and-gas-ternary/dataprep.R [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+
+df_fm <- read.csv("data/FmCodes-Table_1.csv")
+df_fm <- as.data.frame(lapply(df_fm, trimws))
+
+df['fm_name'] <- df['tdfm']
+df_fm <- df_fm[-c(1),] 
+
+formation_name <- list()
+i <- 1
```
Comment (HammadTheOne) [inline apps/dashr-oil-and-gas-ternary/dataprep.R]: I found this interesting, looping using an index counter, I see that a lot in Python but not in R. It works fine, and I think it's more of a stylistic choice, personally I would iteratively loop through the range of elements in R, but I think it's your call how you do it. 👍

**Comment by KPhans** [comment]
@HammadTheOne hmm that is very weird indeed. Ill look into that later today. If all the other options work properly and this one doesn't, then it may not be a problem with the code

**Comment by KPhans** [comment]
@rpkyle @HammadTheOne Fixed the satellite street view. Was a silly filtering where the `:` was changed to a `=` in `"mapbox://styles/mapbox/satellite-streets-v9"`. Since R doesn't utilize colons for equate values. Conflicts have been resolved.

File: apps/dashr-oil-and-gas-ternary/dataprep.R (hunk 11 above)
Comment (KPhans) [reply apps/dashr-oil-and-gas-ternary/dataprep.R]: for now, will stick with an index counter as it seems to work. definitely will use `range` or `seq` later on

**Pull request merged**